source/encoding.c

```diff
@@ -0,0 +1,321 @@
+    }
+
+    tmp /= 3;
+    size_t overflow_check = tmp;
+    tmp = 4 * tmp + 1; /* plus one for the NULL terminator */
+
+    if (AWS_UNLIKELY(tmp < overflow_check)) {
+        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    }
+
+    *encoded_len = tmp;
+
+    return AWS_OP_SUCCESS;
+}
+
+
+int aws_compute_base64_decoded_len(const char *input, size_t len, size_t *decoded_len) {
+    assert(input);
+    assert(decoded_len);
+
+    if (len == 0) {
+        *decoded_len = 0;
+        return AWS_OP_SUCCESS;
+    }
+
+    if (AWS_UNLIKELY(len & 0x03)) {
+        return aws_raise_error(AWS_ERROR_INVALID_BASE64_STR);
+    }
+
+    size_t padding = 0;
+
+    if (len >= 2 && input[len - 1] == '=' && input[len - 2] == '=') { /*last two chars are = */
+        padding = 2;
+    }
+    else if (input[len - 1] == '=') { /*last char is = */
+        padding = 1;
+    }
+
+    size_t tmp = len * 3;
+
+    if (AWS_UNLIKELY(tmp < len)) {
+        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    }
+
+    *decoded_len = (tmp / 4 - padding);
+    return AWS_OP_SUCCESS;
+}
+
+int aws_base64_encode(const uint8_t *restrict to_encode, size_t to_encode_len, char *restrict output, size_t output_size) {
+    assert(to_encode);
+    assert(output);
+
+    size_t encoded_length = 0;
+    if (AWS_UNLIKELY(aws_compute_base64_encoded_len(to_encode_len, &encoded_length))) {
+        return AWS_OP_ERR;
+    }
+
+    if (AWS_UNLIKELY(output_size < encoded_length)) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    size_t buffer_length = to_encode_len;
+    size_t block_count = (buffer_length + 2) / 3;
+    size_t remainder_count = (buffer_length % 3);
+    size_t str_index = 0;
+
+    for (size_t i = 0; i < to_encode_len; i += 3 )
+    {
+        uint32_t block = to_encode[ i ];
+
+        block <<= 8;
+        if (AWS_LIKELY(i + 1 < buffer_length))
+        {
+            block = block | to_encode[ i + 1 ];
+        }
+
+        block <<= 8;
+        if (AWS_LIKELY(i + 2 < to_encode_len))
+        {
+            block = block | to_encode[ i + 2 ];
+        }
+
+        output[str_index++] = base64_encoding_table[(block >> 18) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 12) & 0x3F];
+        output[str_index++] = base64_encoding_table[(block >> 6) & 0x3F];
+        output[str_index++] = base64_encoding_table[block & 0x3F];
+    }
+
+    if(remainder_count > 0)
+    {
+        output[block_count * 4 - 1] = '=';
+        if(remainder_count == 1)
+        {
+            output[block_count * 4 - 2] = '=';
+        }
+    }
+
+    /* it's a string add the null terminator. */
+    output[encoded_length - 1] = 0;
+
+    return AWS_OP_SUCCESS;
+}
+
+static inline int base64_get_decoded_value(char to_decode, uint8_t *value, int8_t allow_sentinal) {
+    if (isalnum(to_decode) || to_decode == '/' || to_decode == '+' || (allow_sentinal && to_decode == '=')) {
```

Comment:
This would be faster if you used a flag value in the decoding table to mark illegal base64 characters rather than this collection of four conditions (including one internal table lookup)

Reply:
Good idea will update